### PR TITLE
Extend type hint

### PR DESCRIPTION
**Pull request by christophmluscher**
@michelwi I think this should be correct?! You agree???

### CODE REVIEW

**Comment by Atticus1806** [comment]
Maybe this can also be moved into the function head. I can't suggest it as change sadly I think, since these lines were not changed by the PR.

**Comment by michelwi** [comment]
> @michelwi I think this should be correct?! You agree???

I agree that the added `rasr.FlowNetwork` is correct here, but I fail to see how the already existing `rasr.FlagDependentFlowAttribute` would have worked without being wrapped in a generic_cache flow or so...

Also there are many more places where this update is applicable in this file.

**Review by Atticus1806 — APPROVED** [review]
Thanks a lot for the large typing changes!

File: meta/system.py [hunk 1]
```diff
@@ -2,24 +2,37 @@
...
+Path = setup_path(__package__)
+
+
+class CorpusObject(tk.Object):
+    """
+    A simple container object to track additional information for a bliss corpus
+    """
+
+    def __init__(self):
+        self.corpus_file = None  # type: Optional[tk.Path] # bliss corpus xml
```
Comment (JackTemaki) [inline meta/system.py:31]:
```suggestion
        self.corpus_file: Optional[tk.Path] = None  # bliss corpus xml
```

File: meta/system.py [hunk 2]
```diff
@@ -2,24 +2,37 @@
...
+        self.audio_dir = None  # type: Optional[tk.Path] # audio directory if paths are relative (usually not needed)
```
Comment (JackTemaki) [inline meta/system.py:32]:
```suggestion
        self.audio_dir: Optional[tk.Path] = None  # audio directory if paths are relative (usually not needed)
```

File: meta/system.py [hunk 3]
```diff
@@ -2,24 +2,37 @@
...
+        self.audio_format = None  # type: Optional[str] # format type of the audio files, see e.g. get_input_node_type()
```
Comment (JackTemaki) [inline meta/system.py:33]:
```suggestion
        self.audio_format: Optional[str] = None  # format type of the audio files, see e.g. get_input_node_type()
```

File: meta/system.py [hunk 4]
```diff
@@ -2,24 +2,37 @@
...
+        self.duration = None  # type: Optional[float] # duration of the corpus, is used to determine job time
```
Comment (JackTemaki) [inline meta/system.py:34]:
```suggestion
        self.duration: Optional[float] = None  # duration of the corpus, is used to determine job time
```

File: meta/system.py [hunk 5]
```diff
@@ -57,16 +71,18 @@ def __init__(self):
         self.scorer_args = {}
         self.scorer_hyp_arg = {}
 
-        self.jobs = {"base": {}}  # type: Dict[str,Dict[str,Job]]
+        self.jobs = {"base": {}}  # type: Dict[str, Dict[str, Union[Job, AlignSplitAccumulateSequence]]]
```
Comment (christophmluscher) [inline meta/system.py:74]:
```suggestion
        self.jobs: Dict[str, Dict[str, Union[Job, AlignSplitAccumulateSequence]]] = {"base": {}}
```

File: meta/system.py [hunk 6]
```diff
@@ -116,12 +132,13 @@ def add_overlay(self, origin, name):
 
         self.jobs[name] = {}
 
-    def copy_from_system(self, origin_system, origin_name, target_name=None):
+    def copy_from_system(self, origin_system: "System", origin_name: str, target_name: Optional[str] = None):
```
Comment (JackTemaki) [inline meta/system.py:135]: Can you try `from __future__ import annotations` to solve the System not being defined problem?

File: meta/system.py [hunk 7]
```diff
@@ -535,12 +565,19 @@ def linear_alignment(self, name, corpus, flow, prefix="", **kwargs):
                 },
             )
 
-    def align(self, name, corpus, flow, feature_scorer, **kwargs):
+    def align(
+        self,
+        name: str,
+        corpus: str,
+        flow: Union[str, List[str], Tuple[str], rasr.FlowNetwork],
+        feature_scorer: Union[str, List[str], Tuple[str], rasr.FeatureScorer],
```
Comment (michelwi) [inline meta/system.py:573]: actually `Tuple[str]` means a Tuple containing a single string. But additionally allowed would be 
`Tuple[str, str]` and `Tuple[str, str, int]`

`List[str]` is also only so-so correct, as only Lists with the elements equivalent to the tuples above would be possible.. Well technically also a list with (str, str, int, Any, Any, ...) would work, as all elements behind list[2] are ignored.. But I don't know a good way to type this..

However in most setups the Tuple version is used in practice I think, so whenever we decide to use typechecking I think the tuples should be typed correctly.

Maybe we should define a `selector_type` to make the union shorter.. i.e.
```
selector_type = Union[str, List, Tuple[str], Tuple[str, str], Tuple[str, str, int]]

[...]

flow : Union[selector_type, rasr.FlowNetwork],
feature_scorer: : Union[selector_type, rasr.FeatureScorer],

**Review by michelwi — APPROVED** [review]

**Review by JackTemaki — APPROVED** [review]

**Pull request merged**